### PR TITLE
GDB: Better Python debugging support

### LINKED ISSUE #20483: install libpython.py when installing python
install libpython.py when installing python

### Rationale

Installing this gdb extension would dramatically improve the debugging experience  with python with C extensions.

### Description

libpython.py is a gdb extension written by the python core developers which provides several helper functions for debugging python c extensions using gdb such as `py-bt` which translates the full C backtrace in to a python readable back trace.  Right now we don't install it. According to the documentation, it is generated when the python library is built.

https://devguide.python.org/gdb/

### Additional information

There are a few other packages which provide gdb extensions such as gcc's libstdc++, so an appealing way to solve this would be to make gdb extendable and have python extend gdb to provide this file, however, this would create a circular dependency between gdb and python since gdb depends on python to provide the embedded interpreter within gdb and this file is meaningless without the gdb and the gdb module.  I'm not sure how to resolve this.

@adamjstewart ping'ing you as requested.  I'm not in a hurry to fix this before the new year.

### General information

- [X] I have run `spack --version` and reported the version of Spack

0.16.0-220-cf37e9276

- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by adamjstewart**
@robertu94 Here's a first draft. `gdb` seems to be able to find the right file without adding anything to `.gdbinit`, but it's crashing for me:
```console
$ gdb
/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/gdb-10.1-vlrhtgaqhz7ntwf6pudrdjzzfzrl3pnq/share/gdb/python/gdb/__init__.py:23: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  from imp import reload
GNU gdb (GDB) 10.1
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-apple-darwin19.6.0".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
(gdb) set debug auto-load
(gdb) file python3
Reading symbols from python3...
warning: `/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/spack-stage-python-3.9.1-rn4ofx2ajnjo5vd5n7352kn244k22qqe/spack-src/spack-build/Programs/python.o': can't open to read symbols: No such file or directory.
(No debugging symbols found in python3)
auto-load: Attempted file "/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/python-3.9.1-rn4ofx2ajnjo5vd5n7352kn244k22qqe/bin/python3.9-gdb.gdb" does not exist.
auto-load: Searching 'set auto-load scripts-directory' path ":${prefix}/share/auto-load".
auto-load: Attempted file "/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/python-3.9.1-rn4ofx2ajnjo5vd5n7352kn244k22qqe/bin/python3.9-gdb.gdb" does not exist.
auto-load: Attempted file "${prefix}/share/auto-load/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/python-3.9.1-rn4ofx2ajnjo5vd5n7352kn244k22qqe/bin/python3.9-gdb.gdb" does not exist.
auto-load: Attempted file "/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/python-3.9.1-rn4ofx2ajnjo5vd5n7352kn244k22qqe/bin/python3.9-gdb.py" exists.
auto-load: Loading python script "/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/python-3.9.1-rn4ofx2ajnjo5vd5n7352kn244k22qqe/bin/python3.9-gdb.py" by extension for objfile "/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/python-3.9.1-rn4ofx2ajnjo5vd5n7352kn244k22qqe/bin/python3.9".
auto-load: Matching file "/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/python-3.9.1-rn4ofx2ajnjo5vd5n7352kn244k22qqe/bin/python3.9-gdb.py" to pattern ""
auto-load: Matched - empty pattern
auto-load: File "/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/python-3.9.1-rn4ofx2ajnjo5vd5n7352kn244k22qqe/bin/python3.9-gdb.py" matches directory "".
/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/spack-stage-gdb-10.1-vlrhtgaqhz7ntwf6pudrdjzzfzrl3pnq/spack-src/gdb/thread.c:95: internal-error: struct thread_info *inferior_thread(): Assertion `current_thread_ != nullptr' failed.
A problem internal to GDB has been detected,
further debugging may prove unreliable.
Quit this debugging session? (y or n) y

This is a bug, please report it.  For instructions, see:
<https://www.gnu.org/software/gdb/bugs/>.

/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/spack-stage-gdb-10.1-vlrhtgaqhz7ntwf6pudrdjzzfzrl3pnq/spack-src/gdb/thread.c:95: internal-error: struct thread_info *inferior_thread(): Assertion `current_thread_ != nullptr' failed.
A problem internal to GDB has been detected,
further debugging may prove unreliable.
Create a core file of GDB? (y or n) y
Abort trap
```
It looks like something still refers to the build stage instead of the installation directory. Adding `+debug` doesn't seem to help. Any ideas?

Closes #20483

### CODE REVIEW

**Comment by robertu94** [comment]
@adamjstewart I’ve seen a bug like this before with some of my own code when used with spack, but I’ll need to investigate to be certain; however I never found a satisfactory fix. In the past, the DWARF (debug symbols) information points to the locations where the files were built rather than where they were installed, and when spack deletes the stage it breaks the debug information. I’ll investigate more this weekend. 

Thanks for the quick prototype on the solution.

**Comment by adamjstewart** [comment]
@skosukhin may be able to help as well

**Comment by robertu94** [comment]
@adamjstewart  I tested this over the weekend, and I didn't get the crash that you did when using Linux, however, I did run into some build issues with GDB relating to a leakage of my system gettext which was built with NLS (native language services);  into the build environment on my laptop and causing a link failure.  I'll probably open a seperate PR for that when I get it working as expected.

When I finally got things building, the safe autoloading didn't work for me, but it did detect the file.  Here is the error message I got:

```console
Reading symbols from python...
warning: File "/home/runderwood/git/contrib/spack/opt/spack/linux-gentoo2-skylake/gcc-10.1.0/python-3.8.6-46imhgqiohzit7d3py2tpnm5quoro7hg/bin/python3.8-gdb.py" auto-loading has been declined by your `auto-load safe-path' set to "$debugdir:$datadir/auto-load".
To enable execution of this file add
        add-auto-load-safe-path /home/runderwood/git/contrib/spack/opt/spack/linux-gentoo2-skylake/gcc-10.1.0/python-3.8.6-46imhgqiohzit7d3py2tpnm5quoro7hg/bin/python3.8-gdb.py
line to your configuration file "/home/runderwood/.gdbinit".
To completely disable this security protection add
        set auto-load safe-path /
line to your configuration file "/home/runderwood/.gdbinit".
For more information about this security protection see the
"Auto-loading safe path" section in the GDB manual.  E.g., run from the shell:
        info "(gdb)Auto-loading safe path"

```

Here are the values that I got for 

`$debugdir` = `/home/runderwood/git/contrib/spack/opt/spack/linux-gentoo2-skylake/gcc-10.1.0/gdb-10.1-dtf7plsp3jvpqx23lkwwatirs2rfzz7f/lib/debug"`
`$datadir` = `"/home/runderwood/git/contrib/spack/opt/spack/linux-gentoo2-skylake/gcc-10.1.0/gdb-10.1-dtf7plsp3jvpqx23lkwwatirs2rfzz7f/share/gdb"`
`auto-load safe-path` is `$debugdir:$datadir/auto-load`  

When I override this in my `~/.gdbinit` to include `/home/runderwood/git/contrib/spack/opt/spack/linux-gentoo2-skylake/gcc-10.1.0/python-3.8.6-46imhgqiohzit7d3py2tpnm5quoro7hg/bin/python3.8-gdb.py` it didn't crash like it did for you, and I get python back traces using `py-bt` like I would expect.  Perhaps the crashes you see are a MacOS bug?

**Comment by adamjstewart** [comment]
Alright, I'll add the auto load logic and we can merge.

**Comment by adamjstewart** [comment]
@robertu94 can you try again with the latest commit?

**Comment by robertu94** [comment]
@adamjstewart that didn't work either when I tested it, so I played around with moving the file and changing its name to get things working.  I found if you put it here named like this:

```console
[runderwood@voyager auto-load]$ pwd
/home/runderwood/git/contrib/spack/opt/spack/linux-gentoo2-skylake/gcc-10.1.0/gdb-10.1-dtf7plsp3jvpqx23lkwwatirs2rfzz7f/share/gdb/auto-load
[runderwood@voyager auto-load]$ tree
.
└── home
    └── runderwood
        └── git
            └── contrib
                └── spack
                    └── opt
                        └── spack
                            └── linux-gentoo2-skylake
                                └── gcc-10.1.0
                                    └── python-3.8.6-46imhgqiohzit7d3py2tpnm5quoro7hg
                                        └── lib
                                            └── libpython3.8d.so.1.0-gdb.py
```

Then it works.  I think the `gdb/share/auto-load` directory expects to have have a path that matches the path to the binary that matches.  On my system, `libpython3.8d.so.1.0` is the name of the python library that spack compiled, and `/home/runderwood/git/contrib/spack/opt/spack/linux-gentoo2-skylake/gcc-10.1.0/python-3.8.6-46imhgqiohzit7d3py2tpnm5quoro7hg/lib` is where it was installed.  The same is also true for `$debug-dir`.

Looking at how the log for the search goes, it looks like gdb checks for 3 possible default locations for gdb scripts for a binary located at `/path/to/binary`

+ `/path/to/binary-gdb.py`
+ `$debug-dir/path/to/binary-gdb.py`
+ `$data-dir/path/to/binary-gdb.py`

This means where the `python3.8-gdb.py` file is placed in spack's python package it also is being detected in addition to the one of the `libpython3.8d.so.1.0-gdb.py` that I copied above. which generates a warning about violating the safe-search path.  However, it then goes on to find the one that I shared above albeit with a warning.

Thanks you for your patience as we debug this.

**Comment by adamjstewart** [comment]
Hmm, we can't put `/home/runderwood/git/contrib/spack` in the path because then the package will no longer be relocatable...

Instead of copying/symlinking the file, can I put a `.gdbinit` anywhere in the `gdb` installation prefix and add Python's `prefix.bin` to the auto-load safe path?

**Comment by robertu94** [comment]
I’ll keep investigating tomorrow, and I’ll let you know what I find.

**Comment by robertu94** [comment]
@adamjstewart there is apparently a directory called `system.gdbinit.d`, however support for it is not compiled by default.  We need to add `--with-system-gdbinit-dir` to the configure arguments.  If the directory listed here is a child of `$prefix` gdb will "relocate" this directory if the installation prefix is copied to a new location (say $install_dir).  For example say it was set to `$prefix/etc/system.gdbinit.d` it would be `$install_dir/etc/system.gdbinit.d`.  These get read before almost everything else.

Here is that I think we should do,  I think we put a python script in this directory which checks the value of the environmental variable `SPACK_ROOT` and adds it to the safe-load path.  Something like this

```python
import gdb
import os

try:
      gdb.execute("add-auto-load-safe-path {0}".format(os.environ['SPACK_ROOT']))
      #using print within python gdb scripts is frowned upon for various reasons, use gdb.write instead
      gdb.write("loaded spack gdb extensions\n")
except:
      gdb.write("failed to load spack gdb extensions\n")

```
As far as I can tell, this file can be named anything and will be read in alphabetical order if they are put in the right directory.  subdirectories are not allowed.

HOWEVER, I had a hard time testing this end-to-end because I couldn't figure out how to to get the package prefix in `configure_args`  I tried `args.append("--with-system-gdbinit-dir="  + self.prefix.etc.join("/system.gdbinit.d"))` but that didn't work.  I 
did test it manually like so, and it loaded the extension without an warning/error.

```console
[runderwood@voyager ~]$ gdb
/home/runderwood/git/contrib/spack/opt/spack/linux-gentoo2-skylake/gcc-10.1.0/gdb-10.1-dtf7plsp3jvpqx23lkwwatirs2rfzz7f/share/gdb/python/gdb/__init__.py:23: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  from imp import reload
GNU gdb (GDB) 10.1
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
*** boost printers: Detecting boost_version... 
*** boost printers: Detected boost version: 1.74.0
*** boost printers: No boost printers are available for boost version 1.74.0!
(gdb) source /tmp/a.py 
loaded spack gdb extensions
(gdb) file /home/runderwood/git/contrib/spack/opt/spack/linux-gentoo2-skylake/gcc-10.1.0/python-3.8.6-46imhgqiohzit7d3py2tpnm5quoro7hg/bin/python
Reading symbols from /home/runderwood/git/contrib/spack/opt/spack/linux-gentoo2-skylake/gcc-10.1.0/python-3.8.6-46imhgqiohzit7d3py2tpnm5quoro7hg/bin/python...
(gdb) break write
Function "write" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (write) pending.
(gdb) run /tmp/b.py
Starting program: /home/runderwood/git/contrib/spack/opt/spack/linux-gentoo2-skylake/gcc-10.1.0/python-3.8.6-46imhgqiohzit7d3py2tpnm5quoro7hg/bin/python /tmp/b.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".

Breakpoint 1, __libc_write (fd=fd@entry=1, buf=buf@entry=0x5555555d6940, nbytes=nbytes@entry=2) at ../sysdeps/unix/sysv/linux/write.c:26
26      ../sysdeps/unix/sysv/linux/write.c: No such file or directory.
(gdb) py-bt
Traceback (most recent call first):
  <built-in method write of _io.TextIOWrapper object at remote 0x7ffff71e1050>
  <built-in method print of module object at remote 0x7ffff7288d70>
  File "/tmp/b.py", line 1, in <module>
    print("a")
(gdb) 
```

In this example, `a.py` contains the python script above, and `b.py` is a testing script.  This is the intended behavior.  The gdb located debugging script located `/home/runderwood/git/contrib/spack/opt/spack/linux-gentoo2-skylake/gcc-10.1.0/python-3.8.6-46imhgqiohzit7d3py2tpnm5quoro7hg/bin/python3.8-gdb.py` without the warning which works and should be fully relocatable.

**Comment by adamjstewart** [comment]
Hmm, this is all rather hacky. I vote we require users to set this themselves in gdbinit and open an issue with gdb asking if there is a better way.

**Comment by robertu94** [comment]
@adamjstewart 

tl;dr the strategy of adding the SPACK_ROOT to auto-load-safe-path at runtime looks like the only way to have this functionality be relocatable and lazy-loadable.  This strategy is similar to what Red Hat based distros do, and what I've seen at least debian based box configured to do, and in my opinion doesn't break either user expectations or the intended security benefits of auto-load-safe-path.  However, if you want to reach out to the GDB devs, I'm happy to wait a while and see what they say.

---

Now that I know what I am looking for, most linux distributions use either `--with-system-gdbinit` or `--with-system-gdbinit-dir` even though `./configure` by default does not use them. 

Fedora, Centos, and RHEL all use `--with-system-gdbinit=/etc/gdbinit` when configuring gdb.  Their system wide gdbinit loops over all files in `/etc/gdbinit.d` with either a `.py` or a `.gdb` extension and sources them like this `gdb.execute('source %s' % f)`.  If we were to adopt a similar strategy, we would be in good company.

Ubuntu and Debian also use `--with-system-gdbinit=/etc/gdb/gdbinit`, however their system wide initialization file is empty allowing the system administrator to make their own changes.  I've seen this file used to in the "real world" to allow debugging scripts installed in a HPC style module loading system.

Gentoo and FreeBSD don't use these options by default, but make it easy to add them and rebuild if the user wants them.

The options `--with-system-gdbinit` or `--with-system-gdbinit-dir` are designed for "site-wide" configuration options, so I think they are designed for this use case: to override default safe directories and set options about where to load debug information from.  If spack were to adopt `debuginfod` -- (a feature in GDB 10+ which allows downloading debug infos and source on the fly) some day in the future, it would be configured in one of these files.

The `safe-autoload` options were designed to indicate which directories contain code that can be "trusted" as safe.  It was designed because a user might be debugging something remotely on an untrusted client machine using gdbserver, and they didn't want a malicious client to provide an auto-loaded gdb file that backdoors the supporting user's machine.  However in the case of spack, we presumably trust other packages build with our own spack instance.

However, we also would like spack packages to be relocatable.  Hard-coding the specific directory at build time won't work for us since it would break this.  There are a few ways to set the auto-load safe path including `--with-auto-load-dir`, but this variable is not relocatable from GDB's perspective meaning if the binary is moved, it won't look in a different location.  This would break relocatability for GDB build spack.

The solution I proposed where we simply set the auto-load-safe path at runtime to include the SPACK_ROOT, or one similar to what Fedora, RHEL, and CentOS where we force-load gdb extensions in certain directories seem like the only ways to fix this.  The benefit of my solution is that by adding SPACK_ROOT to the auto-load-safe-path at runtime, we can load only the modules we need at runtime instead of sourcing them every-time like the Fedora/RHEL/CentOS solution does.

All that said, if you want to reach out to the GDB developers before we go down this route, that's fine by me.  I'm not in a huge hurry to get this merged.

**Comment by adamjstewart** [comment]
Let's set `--with-system-gdbinit=$prefix/etc/gdbinit` and specify the path to the right file. The previous suggestion of using the entire path wasn't relocatable, but specifying it as a string in a file is. I believe we check files for strings containing the installation prefix and replace them during relocation.

**Comment by adamjstewart** [comment]
P.S. Would you like to be listed as an official "maintainer" for the `gdb` package?

**Comment by robertu94** [comment]
I would be happy to be a maintainer for the GDB package.

I use GDB with spack pretty regularly.  I also have a few other ideas for improving the GDB experience with spack, and being the maintainer would put me a position to advocate for these changes.

**Comment by robertu94** [comment]
> Let's set `--with-system-gdbinit=$prefix/etc/gdbinit` and specify the path to the right file. The previous suggestion of using the entire path wasn't relocatable, but specifying it as a string in a file is. I believe we check files for strings containing the installation prefix and replace them during relocation.

If we're going to do this, let's use the `--with-system-gdbinit-dir` option instead or do something like Red Hat did with `--with-system-gdbinit` and we can add files for each package that uses gdb as seperate addtions to the `auto-load-safe-path`.

**Comment by adamjstewart** [comment]
I'm fine with either approach. We can either append to the `gdbinit` file or have separate files for each "extension". Is there any advantage of one over the other?

**Comment by robertu94** [comment]
My thought is that having separate files with names based on the install hash makes it less likely we accidentally clobber something when relocating.  It also setups up GDB to use spack's `extends` feature where that makes sense.

**Comment by adamjstewart** [comment]
We're not using the `extends` feature at the moment, and I don't think it makes sense to do so. `extends` is designed to symlink the entire installation prefix of a package into another package install prefix. This is needed for things like Python libraries, but doesn't seem to be necessary for gdb.

**Comment by robertu94** [comment]
I thought `extend` could also be used to symlink specific files from a different prefix.  Yeah, I agree if it's the entire prefix, we don't need it here.

**Comment by adamjstewart** [comment]
You can select a subset of files to symlink, but I don't think we need to symlink any files here.

**Comment by robertu94** [comment]
It doesn't make sense for _just_ python, but there are a few packages like llvm's `libcxx` and gcc's `libstdc++` which provide pretty printers for the standard library and which are built by llvm and gcc respectively which would all benefit from this functionaity.  Some other packages that provide gdb helpers include `glib`, `gobject`, and `gstreamer`.

Python and GCC's libstdc++ are special cases because gdb depends on these to be built.  Using `extends` as I understand it would create a circular dependency, which would be bad and violate alot of assumptions made in spack about the graph being a dag.

**Comment by adamjstewart** [comment]
Alright, can you try the latest commit?

**Comment by robertu94** [comment]
it works!

**Review by alalazo — APPROVED** [review]
